Fallback to x64 directory if running x64 process on win-arm64 apphost

mateoatr opened the pull request:
Should fix #59473.

Commit:
Append x64 to simulations on win-arm64

mateoatr commented:
@ericstj could you please give this a shot? I did some validation but best I have now is x86 emulation on a win-arm64 machine, we should try the real thing (running x64 on a windows arm64 machine).

jkotas commented on src/native/corehost/hostmisc/pal.windows.cpp:
```diff
@@ -276,8 +276,10 @@ bool pal::get_default_installation_dir(pal::string_t* recv)
+    USHORT pProcessMachine, pNativeMachine;
     const pal::char_t* program_files_dir;
-    if (pal::is_running_in_wow64())
+    auto is_running_in_wow64 = IsWow64Process2(GetCurrentProcess(), &pProcessMachine, &pNativeMachine);
```
`IsWow64Process2` is only present in more recent Win10. Anything older than is missing this API.

You need to get it via LoadLibraryEx / GetProcAddress, and fallback to what was here before if the API is not present.

vitek-karas replied:
On top of @jkotas's comment, the function doesn't return true/false (isWow/isnotWow). The return value is `FALSE` only if the function fails. In which case we used to treat it as if it's not a Wow process, so we should probably continue to do so.

The answer is looking at the `pProcessMachine` which should be "unknown" if it's not a Wow process, otherwise it should be some "known" value. The correct version (if the new API is available) is to check that the `pProcessMachine` is x86, and only then use the `ProgramFiles(x86)` value.

jkotas commented on src/native/corehost/hostmisc/pal.windows.cpp:
```diff
@@ -292,6 +294,14 @@ bool pal::get_default_installation_dir(pal::string_t* recv)
     append_path(recv, _X("dotnet"));
+    if (!is_running_in_wow64 &&
+        pProcessMachine != pNativeMachine &&
+        pNativeMachine == IMAGE_FILE_MACHINE_ARM64)
```
```suggestion
        pProcessMachine == IMAGE_FILE_MACHINE_AMD64)
```
Nit: This would be slightly more future proof.

vitek-karas commented on src/native/corehost/hostmisc/pal.windows.cpp:
```diff
@@ -292,6 +294,14 @@ bool pal::get_default_installation_dir(pal::string_t* recv)
+        pProcessMachine != pNativeMachine &&
+        pNativeMachine == IMAGE_FILE_MACHINE_ARM64)
+    {
+        // We are running an x64 process in a windows arm64 machine.
```
```suggestion
        // We are running an x64 process on a non-x64 windows machine.
```

vitek-karas commented on src/native/corehost/hostmisc/pal.windows.cpp:
```diff
@@ -276,8 +276,10 @@ bool pal::get_default_installation_dir(pal::string_t* recv)
 
+    USHORT pProcessMachine, pNativeMachine;
     const pal::char_t* program_files_dir;
-    if (pal::is_running_in_wow64())
```
The other place where this is called is in utils.cpp `get_dotnet_root_from_env` - we should fix both places to use the same mechanism.

Commit:
PR feedback

Commit:
Update trace and comments

vitek-karas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -350,6 +350,56 @@ bool try_stou(const pal::string_t& str, unsigned* num)
+            return false;
+        }
+
+        auto is_running_in_wow64 = isWow64Process2Func(GetCurrentProcess(), &pProcessMachine, &pNativeMachine);
```
Nit: I would rename this variable - the content is just failure/success, not the actual result.
Maybe just call the function from the if directly and remove the variable.

vitek-karas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -350,6 +350,56 @@ bool try_stou(const pal::string_t& str, unsigned* num)
+        if (!isWow64Process2Func)
+        {
+            // Could not find IsWow64Process2.
+            FreeLibrary(kernel32);
```
Technically this should free the library after every call.

It might be a good idea to actually cache the result of this whole function in a static variable. The result should never change. But it's probably not as important (we would need to make it thread safe BTW).

jkotas replied:
You do not need to bother with freeing kernel32.dll. It is not an unloadable library.

vitek-karas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -358,12 +408,18 @@ bool get_dotnet_root_from_env(pal::string_t* dotnet_root_env_var_name, pal::stri
         if (get_file_path_from_env(dotnet_root_env_var_name->c_str(), recv))
             return true;
     }
+    else if (is_emulating_x64())
+    {
+        *dotnet_root_env_var_name = _X("DOTNET_ROOT_x64");
+        if (get_file_path_from_env(dotnet_root_env_var_name->c_str(), recv))
+            return true;
+    }
```
Why do we need this branch? The function already tries `DOTNET_ROOT_X64` above (since `get_arch` will always be `x64` when `is_emulating_x64` returns true).

jkotas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -350,6 +350,56 @@ bool try_stou(const pal::string_t& str, unsigned* num)
+bool is_emulating_x64()
+{
+    USHORT pProcessMachine, pNativeMachine;
+    auto kernel32 = LoadLibraryEx("kernel32.dll", NULL, 0);
```
```suggestion
    auto kernel32 = LoadLibraryExW(L"kernel32.dll", NULL, LOAD_LIBRARY_SEARCH_SYSTEM32);
```
- Unicode version (W suffix) is faster - avoids ANSI->Unicode conversion
- Use `LOAD_LIBRARY_SEARCH_SYSTEM32` to avoid static analysis warnings about dll hijacking

jkotas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -358,12 +408,18 @@ bool get_dotnet_root_from_env(pal::string_t* dotnet_root_env_var_name, pal::stri
     }
+    else if (is_emulating_x64())
+    {
+        *dotnet_root_env_var_name = _X("DOTNET_ROOT_x64");
```
```suggestion
        *dotnet_root_env_var_name = _X("DOTNET_ROOT_X64");
```
Nit: Match casing from the spec

jkotas commented on src/native/corehost/hostmisc/utils.cpp:
```diff
@@ -358,12 +408,18 @@ bool get_dotnet_root_from_env(pal::string_t* dotnet_root_env_var_name, pal::stri
         if (get_file_path_from_env(dotnet_root_env_var_name->c_str(), recv))
             return true;
     }
+    else if (is_emulating_x64())
```
After closer look, I do not understand why this case is needed at all. It should be taken care of by the `append(to_upper(get_arch()))` case above. 

Do we have a repro or test that demonstrates what this is fixing, and that is failing before this change and passing with this change?

Commit:
Use LoadLibraryExW
No need unloading kernel32 dll
Remove unnecessary variable
Remove unnecessary else-if branch

Commit:
Remove is_running_in_wow64 from unix pal

Commit:
Move is_emulating_x64 to pal
Use sysctlbyname on macOS

Commit:
Update ifdef condition

agocke commented:
@vitek-karas Anything else we need here?
